daemon, config: regenerate endpoint datapath on agent config change

jaffcheng opened the pull request:
Right now changing some agent options like `PolicyAuditMode` from
cilium CLI succeeds but doesn't affect datapath, e.g.:
# cilium config PolicyAuditMode=true
This patch fixes this by applying agent config changes to endpoint
header file and triggering endpoint datapath regeneration.

Fixes: #13902
Signed-off-by: Jaff Cheng <jaff.cheng.sh@gmail.com>

christarazi commented on daemon/cmd/daemon.go:
```diff
@@ -425,6 +428,17 @@ func NewDaemon(ctx context.Context, epMgr *endpointmanager.EndpointManager, dp d
+	if err != nil {
+		return nil, nil, err
+	}
+	d.datapathRegenTrigger = t
```
It seems that we are reusing the same `t` variable. Wouldn't that mess with the variable that's already set to the previous `t`?

jaffcheng replied:
At this point, `t` should already be assigned to `d.policyTrigger` so I think it would be safe to change `t`. I'll use a new name to make it clearer.

christarazi commented on daemon/cmd/daemon.go:
```diff
@@ -803,6 +820,21 @@ func (d *Daemon) TriggerReloadWithoutCompile(reason string) (*sync.WaitGroup, er
+		Reason:            reason,
+		RegenerationLevel: regeneration.RegenerateWithDatapathRewrite,
+	}
+	d.endpointManager.RegenerateAllEndpoints(regenerationMetadata)
```
I believe regenerating all endpoints will already log a msg, do we need to log this again?

christarazi commented on pkg/datapath/linux/config/config.go:
```diff
@@ -400,7 +400,8 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 
-	if option.Config.PolicyAuditMode {
+	val, ok := option.Config.Opts.Opts[option.PolicyAuditMode]
+	if ok && val != option.OptionDisabled {
```
```suggestion
	if val, ok := option.Config.Opts.Opts[option.PolicyAuditMode]; ok && val != option.OptionDisabled {
```

christarazi commented on pkg/endpoint/endpoint.go:
```diff
@@ -684,6 +684,15 @@ func (e *Endpoint) applyOptsLocked(opts option.OptionMap) bool {
 	return changed
 }
 
+func (e *Endpoint) ApplyOpts(opts option.OptionMap) error {
```
Could you add a godoc that describes that this function is?

christarazi commented on pkg/endpointmanager/manager.go:
```diff
@@ -442,6 +442,16 @@ func (mgr *EndpointManager) RegenerateAllEndpoints(regenMetadata *regeneration.E
 	return &wg
 }
 
+func (mgr *EndpointManager) OverrideEndpointOpts(om option.OptionMap) error {
```
Here as well

christarazi reviewed (changes requested):
Thanks for the PR. Overall, I think it looks good. A few items to address below.

christarazi commented on pkg/endpointmanager/manager.go:
```diff
@@ -442,6 +442,17 @@ func (mgr *EndpointManager) RegenerateAllEndpoints(regenMetadata *regeneration.E
+// OverrideEndpointOpts applies the given options to all endpoints.
+func (mgr *EndpointManager) OverrideEndpointOpts(om option.OptionMap) error {
+	eps := mgr.GetEndpoints()
+	for _, ep := range eps {
```
```suggestion
	for _, ep := range mgr.GetEndpoints() {
```

christarazi commented on daemon/cmd/daemon.go:
```diff
@@ -425,6 +428,17 @@ func NewDaemon(ctx context.Context, epMgr *endpointmanager.EndpointManager, dp d
 	d.policyTrigger = t
 
+	regenerationTrigger, err := trigger.NewTrigger(trigger.Parameters{
+		Name:            "datapath_regeneration",
```
```suggestion
		Name:            "datapath-regeneration",
```

There's a pattern in the code to use dashes.

jaffcheng replied:
Ah seems I followed the pattern of `policy_update`
```
./pkg/ipam/node_manager.go-             Name:            "ipam-node-manager-resync",
./pkg/ipam/node_manager.go-                     Name:            fmt.Sprintf("ipam-pool-maintainer-%s", resource.Name),
./pkg/ipam/node_manager.go-                     Name:        fmt.Sprintf("ipam-pool-maintainer-%s-retry", resource.Name),
./pkg/ipam/node_manager.go-                     Name:            fmt.Sprintf("ipam-node-k8s-sync-%s", resource.Name),
./pkg/ipam/crd.go-              Name:        "crd-allocator-node-refresher",
./pkg/endpoint/endpoint.go-                     Name:        "sync_endpoint_header_file",
./daemon/cmd/daemon.go-         Name:            "policy_update",
```

christarazi reviewed (approved):
LGTM, just a few minor nits to address. Doesn't require another look from me. Can you rebase your PR on lastest master to pick up some CI changes? Currently, we have some Docker rate limiting issues and a rebase would pick up the new mitigations, so that we can run tests against this PR.

jibi commented on daemon/cmd/daemon.go:
```diff
@@ -425,6 +428,17 @@ func NewDaemon(ctx context.Context, epMgr *endpointmanager.EndpointManager, dp d
 
+	regenerationTrigger, err := trigger.NewTrigger(trigger.Parameters{
+		Name:            "datapath-regeneration",
+		MetricsObserver: &policyTriggerMetrics{},
```
Should we define a new `MetricsObserver` and a new constant for the `MinInterval` rather then reusing the ones for the policy trigger?

jaffcheng replied:
Not sure if we want to count this in another metric, for now, it's only triggered by agent configuration change, I chose to reuse policyTriggerMetrics according to this comment:
https://github.com/cilium/cilium/blob/9e48678686e213b0b36d616d0bf20f5104eeca04/daemon/cmd/policy.go#L78-L82

christarazi replied:
I think just leaving a small comment to say this is deliberate would be good.

jibi commented on pkg/datapath/linux/config/config.go:
```diff
@@ -400,7 +400,7 @@ func (h *HeaderfileWriter) WriteNodeConfig(w io.Writer, cfg *datapath.LocalNodeC
 	}
 
-	if option.Config.PolicyAuditMode {
+	if val, ok := option.Config.Opts.Opts[option.PolicyAuditMode]; ok && val != option.OptionDisabled {
```
I don't understand this change as that's the only case we are accessing `Config.Opts.Opts` in `config.go`

jaffcheng replied:
Hi, I added a comment to explain this change.

jibi replied:
Thanks! :pray:

jibi replied:
On a second look I think this specific change (using `Config.Opts.Opts[option.PolicyAuditMode]` in place of `option.Config.PolicyAuditMode`) should be moved into a separate commit as it's kind of unrelated to the more general change proposed here of regenerating the datapath program on agent config changes

jaffcheng committed:
datapath: check PolicyAuditMode by mutable config options

PolicyAuditMode can be changed at runtime by cilium CLI/API, so we
should respect the mutable config option.

Fixes: #13902
Signed-off-by: Jaff Cheng <jaff.cheng.sh@gmail.com>

jaffcheng committed:
daemon, config: regenerate endpoint datapath on agent config change

Right now changing some agent options like `PolicyAuditMode` from
cilium CLI succeeds but doesn't affect datapath, e.g.:
  cilium config PolicyAuditMode=true
This patch fixes this by applying agent config changes to endpoint
header file and triggering endpoint datapath regeneration.

Fixes: #13902
Signed-off-by: Jaff Cheng <jaff.cheng.sh@gmail.com>

jibi commented on pkg/endpoint/endpoint.go:
```diff
@@ -681,6 +681,17 @@ func (e *Endpoint) applyOptsLocked(opts option.OptionMap) bool {
+		return false, err
+	}
+	defer e.unlock()
+	changed := e.applyOptsLocked(opts)
```
After giving this another look I have one last question: is it safe to _override_ the endpoint opts? Could this trash any option which was set afterwards and was not part of the ones explicitly stored in `option.Config`?

For example: I was looking at the logic in `policy.go`:
https://github.com/cilium/cilium/blob/f66515219c21041f17d424cb8f2ca2b1b81c4687/pkg/endpoint/policy.go#L298-L312 

which in one specific case is responsible for setting an additional option (`Conntrack`)

jaffcheng replied:
Thanks for the comment, if I understand you correctly, are you concerned about the specific case `opts[option.Conntrack]` not ensured when overriding endpoint opts? I've checked the code path, this is ensured right before regenerating bpf: https://github.com/cilium/cilium/blob/f66515219c21041f17d424cb8f2ca2b1b81c4687/pkg/endpoint/bpf.go#L831

jibi replied:
Thanks for confirming :+1:

jibi reviewed (commented):
I run the unit test locally as Travis CI is unhappy and they are green.
This LGTM now pending one last question

joestringer commented on daemon/cmd/daemon.go:
```diff
@@ -803,6 +823,27 @@ func (d *Daemon) TriggerReloadWithoutCompile(reason string) (*sync.WaitGroup, er
+// This is only called after agent configuration changes for now.
+func (d *Daemon) TriggerDatapathRegen(force bool, reason string) {
+	if force {
+		log.Debugf("Artificially increasing policy revision to enforce policy recalculation")
```
```suggestion
		log.Debug("Artificially increasing policy revision to enforce policy recalculation")
```

joestringer replied:
Would also be good to use a completely different log message here to differentiate it from the one in `TriggerPolicyUpdates()`.

joestringer replied:
Actually, for that matter, does this even need to bump the policy revision? Maybe not.

jaffcheng replied:
> Actually, for that matter, does this even need to bump the policy revision? Maybe not.

I have removed the code bumping policy revision in a previous patch version, but IT failed due to `PolicyEnforcement` config change needs a newer policy revision, so I kept the behavior from `TriggerPolicyUpdates`. I'll add logic to bump policy revision only on changes including `PolicyEnforcement`.

joestringer commented on daemon/cmd/config.go:
```diff
@@ -80,6 +80,11 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
 	changes += option.Config.Opts.ApplyValidated(om, changedOption, d)
+	err = d.endpointManager.OverrideEndpointOpts(om)
+	if err != nil {
+		msg := fmt.Errorf("Override endpoint configuration failed %s", err)
```
```suggestion
		msg := fmt.Errorf("Override endpoint configuration failed %w", err)
```

joestringer commented on pkg/endpointmanager/manager.go:
```diff
@@ -432,6 +432,16 @@ func (mgr *EndpointManager) RegenerateAllEndpoints(regenMetadata *regeneration.E
+func (mgr *EndpointManager) OverrideEndpointOpts(om option.OptionMap) error {
+	for _, ep := range mgr.GetEndpoints() {
+		if _, err := ep.ApplyOpts(om); err != nil {
+			return err
```
If this succeeds for the first endpoint then fails for the second, then any third endpoint will be unaffected. I think that this should probably check the error, then if it is the "endpoint being removed" error, continue, otherwise log the error and continue.

joestringer replied:
^ May need something like this so you have an error to compare correctly:

```
diff --git a/pkg/endpoint/lock.go b/pkg/endpoint/lock.go
index f9f35a1b140a..51c7f0747920 100644
--- a/pkg/endpoint/lock.go
+++ b/pkg/endpoint/lock.go
@@ -16,12 +16,16 @@ package endpoint

 import "fmt"

+var (
+       ErrEndpointDeleted = fmt.Errorf("lock failed: endpoint is in the process of being removed")
+)
+
 // lockAlive returns error if endpoint was removed, locks underlying mutex otherwise
 func (e *Endpoint) lockAlive() error {
        e.mutex.Lock()
        if e.IsDisconnecting() {
                e.mutex.Unlock()
-               return fmt.Errorf("lock failed: endpoint is in the process of being removed")
+               return ErrEndpointDeleted
        }
        return nil
 }
```

joestringer commented on daemon/cmd/config.go:
```diff
@@ -80,6 +80,11 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
+	if err != nil {
+		msg := fmt.Errorf("Override endpoint configuration failed %s", err)
+		return api.Error(PatchConfigFailureCode, msg)
+	}
```
Strictly speaking, this has an unintended impact if some portion of the API request fails: If the above is successful, but then line 95 sees an error.. then the endpoint options will have been configured, but the changes will not have been applied yet. Then the API will return an error, but the changes were already made! Then next time someone tries to regenerate the endpoints for any reason, the changes may take effect.

There's two broad ways to solve this:
1. Undo these changes in the error condition below. One way we have done this in the past is to have the function like `OverrideEndpointOpts()` return a "revert function" (see `revertFunc') which can be run in the error condition below.
1. Move the configuration logic into the per-endpoint eventqueues for modifying endpoint state.

jaffcheng replied:
You are right, agent option changes should be reverted on error as well. I think it'd be better to put this into a separate commit before overriding endpoint opts.

joestringer reviewed (changes requested):
I have some concerns around corner case error condition handling below.

Also some misc nits around Errorf(), Debugf().

christarazi commented on daemon/cmd/daemon.go:
```diff
@@ -225,9 +225,12 @@ func (d *Daemon) init() error {
+			option.Config.ConfigPatchMutex.RUnlock()
 			return err
 		}
+		option.Config.ConfigPatchMutex.RUnlock()
```
Nit: This can be simplified by deferring the `RUnlock()` after the `RLock()`

christarazi commented on daemon/cmd/config.go:
```diff
@@ -82,16 +83,25 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
+		option.Config.ConfigPatchMutex.Unlock()
 		d.TriggerPolicyUpdates(true, "agent configuration update")
+	} else {
+		option.Config.ConfigPatchMutex.Unlock()
```
Same here about deferring the `RUnlock()`. Eliminates the need to have if-else branches. We know that the unlock should take place at the exit points of the function anyway.

christarazi commented on pkg/option/config.go:
```diff
@@ -2136,6 +2136,13 @@ func (c *DaemonConfig) SetIPv4NativeRoutingCIDR(cidr *cidr.CIDR) {
+func (c *DaemonConfig) IPv4NativeRoutingCIDRRLocked() (cidr *cidr.CIDR) {
+	cidr = c.ipv4NativeRoutingCIDR
+	return
+}
```
We can simply return `c.ipv4NativeRoutingCIDR` without needing to make a copy of the pointer. This gets rid of the named return value.

If the intent was to protect `c.ipv4NativeRoutingCIDR`, then a deep copy is required.

christarazi commented on pkg/endpointmanager/manager.go:
```diff
@@ -432,6 +432,17 @@ func (mgr *EndpointManager) RegenerateAllEndpoints(regenMetadata *regeneration.E
+// OverrideEndpointOpts applies the given options to all endpoints.
+func (mgr *EndpointManager) OverrideEndpointOpts(om option.OptionMap) {
+	for _, ep := range mgr.GetEndpoints() {
+		if _, err := ep.ApplyOpts(om); err != nil && err != endpoint.ErrEndpointDeleted {
```
`golint` is advising here to use `errors.Is()` to compare errors.

aanm commented on daemon/cmd/config.go:
```diff
@@ -56,8 +56,13 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
+	// Copy old configurations for potential reversion
 	oldEnforcementValue := policy.GetPolicyEnabled()
 	oldConfigOpts := option.Config.Opts.DeepCopy()
+	oldEpConfigOpts := make(option.OptionMap)
```
small nit `oldEpConfigOpts := make(option.OptionMap, len(om))`

aanm reviewed (approved):
Just one small nit

joestringer commented on daemon/cmd/config.go:
```diff
@@ -51,29 +51,30 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
 
 	// Serialize configuration updates to the daemon.
 	option.Config.ConfigPatchMutex.Lock()
+	defer option.Config.ConfigPatchMutex.Unlock()
```
These locking changes make me nervous. It looks very easy to misunderstand the locking patterns and introduce a deadlock. I've looked for about half an hour now and I can't convince myself it's correct (it may be; but even if it is, it looks easy to break in future).

I believe this is attempting to address the comments here:
https://github.com/cilium/cilium/pull/13971#discussion_r530021699

Furthermore, I think it is trying to address a case like:
1. API request comes in to configure some setting A
   1. Previous configuration is stored
   1. Configuration is applied
   1. Lock is released
   1. Reinitialization is triggered
1. API request comes in to configure some setting B
   1. Configuration is applied concurrently with reinitialization
   1. Lock is released
   1. Reinitialize is triggered
1. The step (1.iv) fails, so we need to undo the attempt at reconfiguration
   1. Overwrite the configuration with the original state, which undoes setting B modifications
1. The step (2.iii) succeeds, but (3) already reverted the changes

Clearly this is not the behaviour we want the API to have so I agree we should address it :+1: 

That said, I don't think it's a good idea to extend the locking over the entire `Reinitialize()` and try to enforce various places underneath to grab the right locks at the right time. This extends the locking over a long period, and also forces various locations to have more understanding of the locking because the locks aren't following a simple obvious "lock, read/change, unlock" pattern. In fact, the `Reinitialize()` path is shelling out to an external script which runs other processes like a compiler and so on, so we have very few guarantees about how long that will take. It could take seconds or even minutes in a CPU-starved environment (not typical, but possible).

The general way we solve this problem in other cases like policyAdd is to have an eventqueue where the events are serialized, then the reconfiguration is handled from there. If the caller wants to wait for the results, there is a channel-based mechanism to listen for the success or failure of the changes. See `putPolicy.Handle` and its callee `Daemon.PolicyAdd()` for examples.

Why is this different from just grabbing the lock? It allows serialization of the API requests without forcing the underlying data structures to be locked for a long period. The locking of the underlying structures, if necessary, can be handled briefly from various goroutines without introducing arbitrarily long delays. I believe this is already the case today for data like `IPv4NativeRoutingCIDR`, which currently has quite granular locking (which this PR is currently proposing to extend to a long-held lock).

With this approach in place, the core logic of a new eventqueue's `struct ConfigModifyEvent` `Handle()` function could store the changes, lock, modify, unlock, `Reinitialize()`, then choose to either Lock+revert+Unlock or return success depending on the result of the reinitialization.

Originally I thought that maybe there's an alternative path, Undoing the changes in the event of a failure without holding the lock the whole time. For simple cases this *might* be doable to check that the settings are configured how we expect before reverting them in the error handling, but overall the most robust approach would be to follow the EventQueue approach.

jaffcheng replied:
Thanks for the detailed write-up! I don't like the locking changes of various places as well. Sorry for not realizing the potential long execution time of `Reinitialize` and the benefit of using an event queue. Will try the suggested approach.

joestringer reviewed (changes requested):
Thanks for following up and investigating the concerns I brought up. I thought a bit longer about the solution here and I think we need to go further to provide strong guarantees and reduce the potential for long blocking mutexes in the agent. More detail below.

jaffcheng committed:
daemon, config: revert agent config changes on error

Currently, if patch agent config API failed to recompile base programs,
API will respond with an error but agent options may have been configured but
not taking effect. These changes may take effect unexpectedly at next regeneration.

This patch reverts agent configuration in such situation.
To make the process of changes and reversion serialized, requests are now
handled in the newly introduced event queue configModifyQueue.

Related: #13902
Signed-off-by: Jaff Cheng <jaff.cheng.sh@gmail.com>

christarazi commented on daemon/cmd/config.go:
```diff
@@ -89,12 +94,53 @@ func (h *patchConfig) Handle(params PatchConfigParams) middleware.Responder {
+	cfgModEvent := eventqueue.NewEvent(c)
+	resChan, err := h.daemon.configModifyQueue.Enqueue(cfgModEvent)
+	if err != nil {
+		msg := fmt.Errorf("enqueue of ConfigModifyEvent failed: %s", err)
```
Nit: `%w` for wrapping errors

christarazi commented on daemon/cmd/daemon.go:
```diff
@@ -92,6 +93,9 @@ import (
 	// AutoCIDR indicates that a CIDR should be allocated
 	AutoCIDR = "auto"
+
+	// Size of the event queue for serializing configuration updates to the daemon
```
Nit: godoc should start with `ConfigModifyQueueSize ...`

joestringer reviewed (approved):
Nice work!
